vizro-core/src/vizro/models/_utils.py

```diff
@@ -0,0 +1,122 @@
+ACTION = "actions"
+
+
+@dataclass
+class PathReplacement:
+    detect_path: str
+    replace_path: str
+    from_import: Callable
+
+
+@dataclass
+class CapturedCallableInfo:
+    name: str
+    module: str
+    args: List[Tuple[str, Any]]
+    code: Optional[str] = None
+
+
+REPLACEMENT_STRINGS = [
+    PathReplacement("plotly.express", "px.", lambda x, y: "import vizro.plotly.express as px"),
+    PathReplacement("vizro.tables", "", lambda x, y: f"from {x} import {y}"),
+    PathReplacement("vizro.figures", "", lambda x, y: f"from {x} import {y}"),
+    PathReplacement("vizro.actions", "", lambda x, y: f"from {x} import {y}"),
+    PathReplacement("vizro.charts", "", lambda x, y: f"from {x} import {y}"),
+]
+
+STANDARD_IMPORT_PATHS = {
+    "import vizro.models as vm",
+    "from vizro import Vizro",
+    "from vizro.managers import data_manager",
+    "from vizro.models.types import capture", #TODO: could make conditional based on content
+}
+
+
+def _get_import_statements(captured_info: List[CapturedCallableInfo]) -> Set[str]:
+    import_paths = set()
+    for info in captured_info:
+        for replacement in REPLACEMENT_STRINGS:
+            if replacement.detect_path in info.module:
+                import_paths.add(replacement.from_import(replacement.detect_path, info.name))
+    return import_paths
+
+
+def _get_callable_code_strings(captured_info: List[CapturedCallableInfo]) -> Set[str]:
+    code_strings = set()
+    for info in captured_info:
+        if info.code is not None:
+            code_strings.add(info.code)
+    return code_strings
+
+
+def _get_data_manager_code_strings(captured_info: List[CapturedCallableInfo]) -> Set[str]:
+    return {
+        f'# data_manager["{arg[1]}"] = ===> Fill in here <==='
+        for info in captured_info
+        for arg in info.args
+        if arg[0] == "data_frame"
+    }
+
+
+def _clean_module_string(module_string: str) -> str:
+    return next(
+        (replacement.replace_path for replacement in REPLACEMENT_STRINGS if replacement.detect_path in module_string),
+        "",
+    )
+
+
+def _repr_clean(info: CapturedCallableInfo) -> str:
+    """Alternative __repr__ method with cleaned module paths."""
+    args = ", ".join(f"{key}={value!r}" for key, value in info.args)
+    module_path = f"{info.module}"
+    modified_module_path = _clean_module_string(module_path)
+    x = f"{modified_module_path}{info.name}({args})"
+    return x
+
+
+def _dict_to_python(model_data: Any, captured_info: Optional[List[CapturedCallableInfo]] = None) -> str:
+    """Function to generate python string from pydantic model dict."""
+    from vizro.models.types import CapturedCallable  # TODO: can we get rid of this?
+
+    if captured_info is None:
+        captured_info = []
+
+    if isinstance(model_data, Dict):
+        if MODEL_NAME in model_data:
+            model_name = model_data.pop(MODEL_NAME)
+            if model_name == ACTIONS_CHAIN:
+                action_data = model_data[ACTION]
```

Comment:
My guide is always - if used twice, put it in a constant. But then I felt the other strings looked lonely, so I converted them too :)

Reply:
This still confused me upon re-reading the code fresh a couple of weeks after I first made this comment.

IMO the problem with the constants is that they just introduce an unnecessary indirection since when I read the code I have to click elsewhere to find out what `ACTION` etc. are rather than it just being there directly in the code. Especially when `ACTION` is literally just `"action"`.

Happy to see what a new reviewer thinks though if you disagree 🙂